docs/docsgen/source/technical/int4.md

```diff
@@ -0,0 +1,55 @@
+<!--
+Copyright (c) ONNX Project Contributors
+
+SPDX-License-Identifier: Apache-2.0
+-->
+
+(onnx-detail-int4)=
+
+# 4 bit integer types
+
+## Papers
+
+Several papers have been published in 2023 to introduce 4 bit integers and their usage in LLMs. Although their range is
+limited, with careful selection of scaling parameters, good accuracy is obtained when used for compression of weights
+(weight-only quantization), and in some cases for quantization of activations as well.
+
+[AWQ: Activation-aware Weight Quantization for LLM Compression and Acceleration](https://arxiv.org/abs/2306.00978)
+Activation-aware Weight Quantization (AWQ) focuses on the quantization of weights in LLMs by considering the
+observation that not all weights are equally important. The method aims to protect salient weights based on the
+activation, rather than relying on backpropagation or reconstruction techniques. By searching for the optimal
+per-channel scaling that preserves the crucial weights, AWQ aims to minimize quantization errors.
+
+[GPTQ: Accurate Post-Training Quantization for Generative Pre-trained Transformers](https://arxiv.org/abs/2210.17323)
+GPTQ proposes a one-shot weight quantization method based on approximate second-order information. GPTQ achieves
+significant compression gains, reducing the bit-width to 3 or 4 bits per weight with negligible accuracy degradation
+compared to the uncompressed baseline.
+
+[Understanding INT4 Quantization for Transformer Models: Latency Speedup, Composability, and Failure Cases](https://arxiv.org/abs/2301.12017)
+This paper discusses quantization of both weights and activations to 4 bit (W4A4). Results indicate that W4A4
+quantization leads to little to no accuracy degradation for encoder-only and encoder-decoder models but results in
+a significant accuracy drop for decoder-only models. To realize the performance gains using W4A4, the study introduces
+a highly optimized end-to-end W4A4 encoder inference pipeline that supports various quantization strategies.
+
+As a result, two new types were introduced in `onnx==1.17.0` supporting a limited set of operators to enable compression using
+4 bit data-types:
+- `UINT4`: 4 bit unsigned integer, values in range [0, 15]
+- `INT4`: 4 bit signed integer, using two's complement represntation. Values in range [-8, 7].
+
+## Cast
+
+Cast from 4 bit to any higher precision type is exact.
+Cast to a 4 bit type is done by rounding to the nearest-integer (with ties to even)
+nearest-even integer and truncating.
+
+## Packing and Unpacking
+
+All 4 bit types are stored as 2x4bit in a single byte.
+The first element is stored in the 4 LSB and the second element is stored in the 4 MSB.
```

Comment:
> The first element is stored in the 4 LSB and the second element is stored in the 4 MSB.

Would it be possible to provide references on the decision on which element is in LSB and which is in the MSB? Do we have a survey on the convention and pros and cons? Is this consistent with the ONNX standard conventions?

Reply:
I think the main consistency check should be with popular models out there that use int4 (and frameworks like pytorch, if they support it). Ideally, we should be able to reuse serialized multi-gigabyte weight tensors produced by existing frameworks/models without having to swap elements within a byte.

Reply:
Torch quint4x2 data type uses little-endian, and this was adopted both by Intel's [Neural Compressor](https://github.com/intel/neural-compressor) and NVIDIA's [TensorRT LLM](https://github.com/NVIDIA/TensorRT-LLM).
Searching for int4 models in HuggingFace shows either little endian is used, or packing method is not specified. 
Another popular implementation for packing int4 [chatglm](https://huggingface.co/THUDM/chatglm-6b-int4/blob/main/quantization_kernels.c)

I did not find any framework or model using big-endian packing of weights.

Reply:
Great - and this is consistent with ONNX preferring little endian in general. Do you think it would be helpful to add what to linked to above as references in this text?

Reply:
I would have mentioned it if I could cite a well established standard or a paper on this, or even if this was stated in the documentation of the open source libraries I've mentioned. However, I had to dig into the source code to find the packing order. Linking to code of an open source library seems a little unorthodox.
I hope that stating this clearly in the ONNX documentation would cause others to follow suit.